{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "59be1fb3-4409-4771-bb76-861ce91eab7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Verify necessary packages installed\n",
    "# !pip install transformers evaluate datasets requests pandas xlrd scikit-learn torch tensorflow openpyxl\n",
    "\n",
    "import pandas as pd, transformers, openpyxl, random\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ed743ba0-633f-4baf-a556-56f497949c2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Select a model checkpoint. For my purposes, training on 3080 Ti at home with 12GB RAM\n",
    "#Will select smaller model because I don't want to overburden my GPU =)\n",
    "#Models\n",
    "#Checkpoint name\tNum layers\tNum parameters\n",
    "#esm2_t48_15B_UR50D\t48\t15B\n",
    "#esm2_t36_3B_UR50D\t36\t3B\n",
    "#esm2_t33_650M_UR50D\t33\t650M\n",
    "#esm2_t30_150M_UR50D\t30\t150M\n",
    "#esm2_t12_35M_UR50D\t12\t35M\n",
    "#esm2_t6_8M_UR50D\t6\t8M\n",
    "\n",
    "model_checkpoint = \"facebook/esm2_t12_35M_UR50D\"\n",
    "\n",
    "mass_spec_data = \"../data/NIHMS1541512-supplement-Sup_Data1.xlsx\"\n",
    "\n",
    "filter_cell_line_peptides = \"../data/NIHMS1541512-supplement-Sup_Tab1.xlsx\" #table e\n",
    "filter_patient_peptides = \"../data/NIHMS1541512-supplement-Sup_Tab6.xlsx\" #table a\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5b5fab0c-4d9c-4cc0-80db-43b81179a3b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create peptide filtration dict\n",
    "#Actually, these look like filtered-in data points, based on Suppl. 1d metrics\n",
    "filtration_dict = {}\n",
    "\n",
    "filter_in_df = pd.read_excel(filter_cell_line_peptides, engine=\"openpyxl\", sheet_name=\"e. Filtered peptide list\")\n",
    "alleles = list(filter_in_df['Allele'].drop_duplicates())\n",
    "\n",
    "#populate sets\n",
    "for allele in alleles:\n",
    "    filtration_dict[allele] = set(filter_in_df.query(\"Allele == @allele\")['Peptide'])\n",
    "\n",
    "\n",
    "df2 = pd.read_excel(filter_patient_peptides, engine=\"openpyxl\", sheet_name=\"a. Filtered peptide list\") \n",
    "\n",
    "#filter patient peptides ubiquitously \n",
    "filtration_dict[\"patient_data\"] = set(df2['Peptide'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96ec4654-7663-472f-8359-2049dd789737",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # This was operating under the assumption that the filtration list was filter-out,\n",
    "# # when it is in fact filter-in\n",
    "# retained_neoepitopes={}\n",
    "# filtered_counts={} \n",
    "# for allele in alleles:\n",
    "#     allele_df = pd.read_excel(mass_spec_data, engine=\"openpyxl\", sheet_name=allele)\n",
    "#     print(len(allele_df))\n",
    "#     filtered_allele_df = allele_df.query(\n",
    "#         \"sequence not in @filtration_dict[@allele]\")# and sequence not in @filtration_dict['patient_data']\") #This is just monoallelic data\n",
    "#     filtered_counts[allele] = len(allele_df) - len(filtered_allele_df)\n",
    "#     print(f\"{str(filtered_counts[allele])} neoepitopes filtered from {allele}; {len(filtered_allele_df)} retained.\")\n",
    "#     retained_neoepitopes[allele] = set(filtered_allele_df['sequence'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "09f8e77a-8477-48a1-9c7c-c26fb0ac5a7f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allele</th>\n",
       "      <th>Peptide</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Length</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>5177</td>\n",
       "      <td>5177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2541</td>\n",
       "      <td>2541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1027</td>\n",
       "      <td>1027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>590</td>\n",
       "      <td>590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>433</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>377</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>294</td>\n",
       "      <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>241</td>\n",
       "      <td>241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>173</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>172</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>159</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>122</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>101</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>47</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Allele  Peptide\n",
       "Length                 \n",
       "12        5177     5177\n",
       "13        2541     2541\n",
       "14        1027     1027\n",
       "15         590      590\n",
       "16         433      433\n",
       "17         377      377\n",
       "18         294      294\n",
       "19         241      241\n",
       "20         173      173\n",
       "21         172      172\n",
       "22         114      114\n",
       "23         159      159\n",
       "24         143      143\n",
       "25         122      122\n",
       "26         101      101\n",
       "27         110      110\n",
       "28          93       93\n",
       "29          68       68\n",
       "30          54       54\n",
       "31          64       64\n",
       "32          57       57\n",
       "33          48       48\n",
       "34          47       47\n",
       "35          26       26\n",
       "36          28       28\n",
       "37           8        8\n",
       "38           7        7\n",
       "39           2        2\n",
       "41           1        1\n",
       "42           2        2"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#This is already in the correct format. It's always fun to waste effort!\n",
    "#For simplicity, will exclude HLA-G\n",
    "filter_in_df = filter_in_df.loc[~filter_in_df.Allele.apply(lambda x: x.startswith('G'))].copy()\n",
    "filter_in_df.loc[filter_in_df.Length > 11].groupby('Length').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "22a69042-0a38-45e0-bd73-b87c8170132d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get size of negatives we need to sample\n",
    "sample_sizes = (filter_in_df.groupby(\"Length\").count()['Allele']*1000).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dd44b584-1afa-487a-a3a1-59218735765c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "from Bio import SeqIO\n",
    "\n",
    "#Get fasta files, parse for alleles of interest\n",
    "#For simplicity of demonstration, will not include HLA-G\n",
    "hla_a_file=\"https://raw.githubusercontent.com/ANHIG/IMGTHLA/Latest/fasta/A_prot.fasta\"\n",
    "hla_b_file=\"https://raw.githubusercontent.com/ANHIG/IMGTHLA/Latest/fasta/B_prot.fasta\"\n",
    "hla_c_file=\"https://raw.githubusercontent.com/ANHIG/IMGTHLA/Latest/fasta/C_prot.fasta\"\n",
    "#hla_g_file=\"https://raw.githubusercontent.com/ANHIG/IMGTHLA/Latest/fasta/G_prot.fasta\"\n",
    "\n",
    "allele_mapper = {}\n",
    "\n",
    "hla_files = [hla_a_file, hla_b_file, hla_c_file]#, hla_g_file]\n",
    "#Loop through fasta files, get record of each HLA and add it to mapper if not present\n",
    "for hla_file in hla_files:\n",
    "    output_file = hla_file.split('/')[-1]\n",
    "    write_dir = f\"../data/{output_file}\"\n",
    "    urllib.request.urlretrieve(hla_file, write_dir)\n",
    "    fasta_sequences = SeqIO.parse(open(write_dir),'fasta')\n",
    "    for fasta in fasta_sequences:\n",
    "        name, sequence = fasta.id, str(fasta.seq)\n",
    "        description = fasta.description\n",
    "        allele = description.split(\" \")[1]\n",
    "\n",
    "        allele_name = ''.join(allele.split(':')[:2]).replace('*', '')\n",
    "        allele_suffix = ''.join(allele.split(':')[2:])\n",
    "\n",
    "#        if ((allele_suffix in ['0101', '01', '']) and (allele_name not in allele_mapper)):\n",
    "        #Naively assume that first entry is best entry. I'm sure this'll come back to bite me\n",
    "        if ((allele_name in list(filter_in_df['Allele'].drop_duplicates())) and\n",
    "            (allele_name not in allele_mapper)): \n",
    "            allele_mapper[allele_name] = sequence\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "assert(len(allele_mapper) == len(filter_in_df.Allele.drop_duplicates())) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a2b445c8-16c7-43b5-bc0d-369095c4d501",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20654it [10:41:22,  1.86s/it]\n"
     ]
    }
   ],
   "source": [
    "#Get human proteome for negative sampling\n",
    "import gzip \n",
    "import random\n",
    "from tqdm import tqdm\n",
    "\n",
    "#TODO: REDO THIS CELL, TAKES TOO LONG TO RUN\n",
    "\n",
    "def get_random_str(main_str, substr_len):\n",
    "    #https://stackoverflow.com/questions/58334846/randomly-extracting-substrings-of-equal-length-from-the-main-string\n",
    "    idx = random.randrange(0, len(main_str) - substr_len + 1)    # Randomly select an \"idx\" such that \"idx + substr_len <= len(main_str)\".\n",
    "    return main_str[idx : (idx+substr_len)]\n",
    "\n",
    "proteome_link = \"https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/UP000005640_9606.fasta.gz\"\n",
    "urllib.request.urlretrieve(proteome_link,\n",
    "                           '../data/UP000005640_9606.fasta.gz') \n",
    "\n",
    "#Determine length of proteome\n",
    "proteome_length = 0\n",
    "with gzip.open('../data/UP000005640_9606.fasta.gz', 'rt') as handle:\n",
    "    proteine_sequences = SeqIO.parse(handle, 'fasta')\n",
    "    for fasta in proteine_sequences:\n",
    "        proteome_length +=1\n",
    "\n",
    "#Create 1000 negatives for every positive\n",
    "#This is very loopy and is expected to take a while...\n",
    "proteome_dict = {}\n",
    "with gzip.open('../data/UP000005640_9606.fasta.gz', 'rt') as handle:\n",
    "    protein_sequences = SeqIO.parse(handle, 'fasta')\n",
    "    for i, fasta in enumerate(tqdm(protein_sequences)):\n",
    "        for sample_length, sample_size in sample_sizes.items(): #Determine how many samples needed from each protein\n",
    "            if sample_length not in proteome_dict:\n",
    "                proteome_dict[sample_length] = set()\n",
    "            name, sequence = fasta.id, str(fasta.seq)\n",
    "            if len(proteome_dict[sample_length]) >= sample_size:\n",
    "                continue #no samples needed here\n",
    "            #If < 1 needed from each, just sample 3 until we have enough\n",
    "            samples_needed_from_sequence = int(max((sample_size/proteome_length), 3))\n",
    "            #Verify no overlap with binding alleles, conservatively\n",
    "            new_protein_set = set()\n",
    "            for sample in range(0, samples_needed_from_sequence):\n",
    "                try:\n",
    "                    new_protein_set.add(get_random_str(sequence, sample_length))\n",
    "                except:\n",
    "                    break #Just move on, not worth resolving for a negative sampler\n",
    "\n",
    "            new_protein_set = new_protein_set - set(filter_in_df['Peptide'])            \n",
    "            proteome_dict[sample_length] = new_protein_set.union(proteome_dict[sample_length])\n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "87e1ab89-25ee-4d3c-a4b0-6575ac5d9f7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pickle\n",
    "\n",
    "# #dump pickle so I don't have to do previous step again (for now)\n",
    "# with open(\"../data/negative_samples.pkl\", 'wb') as f:\n",
    "#     pickle.dump(proteome_dict, f)\n",
    "\n",
    "# backup_df = filter_in_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "74596ac9-697d-410b-8385-f1b2d1ee8870",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open (\"../data/negative_samples.pkl\", \"rb\") as f:\n",
    "    proteome_dict = pickle.load(f)\n",
    "\n",
    "backup_df = filter_in_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "86a6aaf8-32e3-4dba-8871-2eca6a6318ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████| 92/92 [03:15<00:00,  2.13s/it]\n"
     ]
    }
   ],
   "source": [
    "filter_in_df = backup_df.copy()\n",
    "\n",
    "\n",
    "negs = 5\n",
    "filter_in_df['label'] = 1\n",
    "\n",
    "#Create `negs` negatives for every positive\n",
    "for i, allele in enumerate(tqdm(list(filter_in_df.Allele.drop_duplicates()))):\n",
    "    sub_df = filter_in_df.query('Allele == @allele').copy()\n",
    "    for length in list(sub_df['Length'].drop_duplicates()):\n",
    "        sub_df_by_len = sub_df.query('Length == @length')\n",
    "        use_int = negs * len(sub_df_by_len)\n",
    "        add_these = pd.DataFrame({'Allele': [allele]*use_int,\n",
    "                     'Length': [length]*use_int, \n",
    "                     'Peptide': random.sample(proteome_dict[length], use_int),\n",
    "                     'label': [0]*use_int})\n",
    "        filter_in_df = filter_in_df.append(add_these, ignore_index=True)\n",
    "\n",
    "filter_in_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "filter_in_df['allele_sequence'] = filter_in_df['Allele'].map(allele_mapper)\n",
    "filter_in_df['predict_on'] = filter_in_df['allele_sequence'] + filter_in_df['Peptide'] \n",
    "filter_in_df['predict_on_len'] = filter_in_df['predict_on'].apply(lambda x: len(x))\n",
    "\n",
    "                     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b53c1a84-573b-4a87-96ae-fb83d8c66ac4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allele</th>\n",
       "      <th>Length</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>label</th>\n",
       "      <th>allele_sequence</th>\n",
       "      <th>predict_on</th>\n",
       "      <th>predict_on_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0101</td>\n",
       "      <td>8</td>\n",
       "      <td>ADMGHLKY</td>\n",
       "      <td>1</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0101</td>\n",
       "      <td>8</td>\n",
       "      <td>ELDDTLKY</td>\n",
       "      <td>1</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0101</td>\n",
       "      <td>8</td>\n",
       "      <td>FSDNIEFY</td>\n",
       "      <td>1</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0101</td>\n",
       "      <td>8</td>\n",
       "      <td>FTELAILY</td>\n",
       "      <td>1</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0101</td>\n",
       "      <td>8</td>\n",
       "      <td>GLDEPLLK</td>\n",
       "      <td>1</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105093</th>\n",
       "      <td>C1701</td>\n",
       "      <td>39</td>\n",
       "      <td>ALELVTEKGHTFAEELQKIQCTLQDVGSALATPCSSARE</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105094</th>\n",
       "      <td>C1701</td>\n",
       "      <td>39</td>\n",
       "      <td>LKDLEKWQNNLLPSRQFGFIVLTTSAGIMDHEEARRKHT</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105095</th>\n",
       "      <td>C1701</td>\n",
       "      <td>39</td>\n",
       "      <td>QKNFLSYDCGSDKVLSMGHLEEQLYATDAWGKQLEMLRE</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105096</th>\n",
       "      <td>C1701</td>\n",
       "      <td>39</td>\n",
       "      <td>PEEKGNIVEEQPQKNTTEKLGVSAPTIPVRRRLAWDTEN</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105097</th>\n",
       "      <td>C1701</td>\n",
       "      <td>39</td>\n",
       "      <td>SVHLQSETKSIDTESQGSSHFSPPDDIGSAESNCNYKDG</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>411</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1105098 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Allele  Length                                  Peptide  label  \\\n",
       "0        A0101       8                                 ADMGHLKY      1   \n",
       "1        A0101       8                                 ELDDTLKY      1   \n",
       "2        A0101       8                                 FSDNIEFY      1   \n",
       "3        A0101       8                                 FTELAILY      1   \n",
       "4        A0101       8                                 GLDEPLLK      1   \n",
       "...        ...     ...                                      ...    ...   \n",
       "1105093  C1701      39  ALELVTEKGHTFAEELQKIQCTLQDVGSALATPCSSARE      0   \n",
       "1105094  C1701      39  LKDLEKWQNNLLPSRQFGFIVLTTSAGIMDHEEARRKHT      0   \n",
       "1105095  C1701      39  QKNFLSYDCGSDKVLSMGHLEEQLYATDAWGKQLEMLRE      0   \n",
       "1105096  C1701      39  PEEKGNIVEEQPQKNTTEKLGVSAPTIPVRRRLAWDTEN      0   \n",
       "1105097  C1701      39  SVHLQSETKSIDTESQGSSHFSPPDDIGSAESNCNYKDG      0   \n",
       "\n",
       "                                           allele_sequence  \\\n",
       "0        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "1        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "2        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "3        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "4        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "...                                                    ...   \n",
       "1105093  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105094  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105095  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105096  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105097  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "\n",
       "                                                predict_on  predict_on_len  \n",
       "0        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             373  \n",
       "1        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             373  \n",
       "2        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             373  \n",
       "3        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             373  \n",
       "4        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             373  \n",
       "...                                                    ...             ...  \n",
       "1105093  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             411  \n",
       "1105094  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             411  \n",
       "1105095  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             411  \n",
       "1105096  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             411  \n",
       "1105097  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             411  \n",
       "\n",
       "[1105098 rows x 7 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# #dump pickle so I don't have to do previous step again (for now)\n",
    "# with open(\"../data/training_samples_5prev.pkl\", 'wb') as f:\n",
    "#     pickle.dump(filter_in_df, f)\n",
    "\n",
    "# filter_in_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "12f8aede-f796-4bc0-bb29-5d9b978c8b7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open (\"../data/training_samples_5prev.pkl\", \"rb\") as f:\n",
    "    filter_in_df = pickle.load(f)\n",
    "\n",
    "backup_df = filter_in_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0c14091d-dc1e-4c68-92c0-97eaf646f8c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████| 92/92 [00:08<00:00, 11.37it/s]\n"
     ]
    }
   ],
   "source": [
    "#Need training data to have data points from all alleles!\n",
    "filter_in_df.sort_values('Allele', inplace=True)\n",
    "filter_in_df.reset_index(drop=True,inplace=True)\n",
    "filter_in_df['training']=1 #training == 1, test == 2\n",
    "for i, allele in enumerate(tqdm(list(filter_in_df.Allele.drop_duplicates()))):\n",
    "    filter_in_df.loc[filter_in_df.Allele == allele,'training'] = filter_in_df.loc[\n",
    "    filter_in_df.Allele == allele,'training'].apply(lambda x: x + random.choices([0, 1], weights=(75, 25))[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "eaef4c0a-6dc5-4181-a302-ee456dc918b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "829159\n",
      "275939\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allele</th>\n",
       "      <th>Length</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>label</th>\n",
       "      <th>allele_sequence</th>\n",
       "      <th>predict_on</th>\n",
       "      <th>predict_on_len</th>\n",
       "      <th>training</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0101</td>\n",
       "      <td>8</td>\n",
       "      <td>ADMGHLKY</td>\n",
       "      <td>1</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>373</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0101</td>\n",
       "      <td>11</td>\n",
       "      <td>PEKIAVEIPKR</td>\n",
       "      <td>0</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>376</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0101</td>\n",
       "      <td>9</td>\n",
       "      <td>GHQVALSSI</td>\n",
       "      <td>0</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>374</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0101</td>\n",
       "      <td>11</td>\n",
       "      <td>KQKEVFLPSTP</td>\n",
       "      <td>0</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>376</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0101</td>\n",
       "      <td>11</td>\n",
       "      <td>LLSNSSSLWRS</td>\n",
       "      <td>0</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>376</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105093</th>\n",
       "      <td>C1701</td>\n",
       "      <td>9</td>\n",
       "      <td>HSVPPVSRK</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>381</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105094</th>\n",
       "      <td>C1701</td>\n",
       "      <td>9</td>\n",
       "      <td>HPTDPTVLI</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>381</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105095</th>\n",
       "      <td>C1701</td>\n",
       "      <td>9</td>\n",
       "      <td>KKKVVFCPV</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>381</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105096</th>\n",
       "      <td>C1701</td>\n",
       "      <td>9</td>\n",
       "      <td>SEADPQALL</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>381</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105097</th>\n",
       "      <td>C1701</td>\n",
       "      <td>39</td>\n",
       "      <td>SVHLQSETKSIDTESQGSSHFSPPDDIGSAESNCNYKDG</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>411</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1105098 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Allele  Length                                  Peptide  label  \\\n",
       "0        A0101       8                                 ADMGHLKY      1   \n",
       "1        A0101      11                              PEKIAVEIPKR      0   \n",
       "2        A0101       9                                GHQVALSSI      0   \n",
       "3        A0101      11                              KQKEVFLPSTP      0   \n",
       "4        A0101      11                              LLSNSSSLWRS      0   \n",
       "...        ...     ...                                      ...    ...   \n",
       "1105093  C1701       9                                HSVPPVSRK      0   \n",
       "1105094  C1701       9                                HPTDPTVLI      0   \n",
       "1105095  C1701       9                                KKKVVFCPV      0   \n",
       "1105096  C1701       9                                SEADPQALL      0   \n",
       "1105097  C1701      39  SVHLQSETKSIDTESQGSSHFSPPDDIGSAESNCNYKDG      0   \n",
       "\n",
       "                                           allele_sequence  \\\n",
       "0        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "1        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "2        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "3        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "4        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "...                                                    ...   \n",
       "1105093  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105094  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105095  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105096  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105097  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "\n",
       "                                                predict_on  predict_on_len  \\\n",
       "0        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             373   \n",
       "1        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             376   \n",
       "2        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             374   \n",
       "3        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             376   \n",
       "4        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             376   \n",
       "...                                                    ...             ...   \n",
       "1105093  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             381   \n",
       "1105094  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             381   \n",
       "1105095  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             381   \n",
       "1105096  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             381   \n",
       "1105097  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             411   \n",
       "\n",
       "         training  \n",
       "0               1  \n",
       "1               2  \n",
       "2               1  \n",
       "3               1  \n",
       "4               1  \n",
       "...           ...  \n",
       "1105093         2  \n",
       "1105094         1  \n",
       "1105095         1  \n",
       "1105096         1  \n",
       "1105097         2  \n",
       "\n",
       "[1105098 rows x 8 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(filter_in_df.loc[filter_in_df.training==1]))\n",
    "print(len(filter_in_df.loc[filter_in_df.training==2]))\n",
    "filter_in_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c35ab034-fcc3-44e4-bc8f-43072eafd9d1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allele</th>\n",
       "      <th>Length</th>\n",
       "      <th>Peptide</th>\n",
       "      <th>label</th>\n",
       "      <th>allele_sequence</th>\n",
       "      <th>predict_on</th>\n",
       "      <th>predict_on_len</th>\n",
       "      <th>training</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0101</td>\n",
       "      <td>8</td>\n",
       "      <td>ADMGHLKY</td>\n",
       "      <td>1</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>373</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0101</td>\n",
       "      <td>11</td>\n",
       "      <td>PEKIAVEIPKR</td>\n",
       "      <td>0</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>376</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0101</td>\n",
       "      <td>9</td>\n",
       "      <td>GHQVALSSI</td>\n",
       "      <td>0</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>374</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0101</td>\n",
       "      <td>11</td>\n",
       "      <td>KQKEVFLPSTP</td>\n",
       "      <td>0</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>376</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0101</td>\n",
       "      <td>11</td>\n",
       "      <td>LLSNSSSLWRS</td>\n",
       "      <td>0</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "      <td>376</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105093</th>\n",
       "      <td>C1701</td>\n",
       "      <td>9</td>\n",
       "      <td>HSVPPVSRK</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>381</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105094</th>\n",
       "      <td>C1701</td>\n",
       "      <td>9</td>\n",
       "      <td>HPTDPTVLI</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>381</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105095</th>\n",
       "      <td>C1701</td>\n",
       "      <td>9</td>\n",
       "      <td>KKKVVFCPV</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>381</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105096</th>\n",
       "      <td>C1701</td>\n",
       "      <td>9</td>\n",
       "      <td>SEADPQALL</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>381</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105097</th>\n",
       "      <td>C1701</td>\n",
       "      <td>39</td>\n",
       "      <td>SVHLQSETKSIDTESQGSSHFSPPDDIGSAESNCNYKDG</td>\n",
       "      <td>0</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...</td>\n",
       "      <td>411</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1105098 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Allele  Length                                  Peptide  label  \\\n",
       "0        A0101       8                                 ADMGHLKY      1   \n",
       "1        A0101      11                              PEKIAVEIPKR      0   \n",
       "2        A0101       9                                GHQVALSSI      0   \n",
       "3        A0101      11                              KQKEVFLPSTP      0   \n",
       "4        A0101      11                              LLSNSSSLWRS      0   \n",
       "...        ...     ...                                      ...    ...   \n",
       "1105093  C1701       9                                HSVPPVSRK      0   \n",
       "1105094  C1701       9                                HPTDPTVLI      0   \n",
       "1105095  C1701       9                                KKKVVFCPV      0   \n",
       "1105096  C1701       9                                SEADPQALL      0   \n",
       "1105097  C1701      39  SVHLQSETKSIDTESQGSSHFSPPDDIGSAESNCNYKDG      0   \n",
       "\n",
       "                                           allele_sequence  \\\n",
       "0        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "1        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "2        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "3        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "4        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...   \n",
       "...                                                    ...   \n",
       "1105093  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105094  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105095  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105096  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "1105097  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...   \n",
       "\n",
       "                                                predict_on  predict_on_len  \\\n",
       "0        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             373   \n",
       "1        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             376   \n",
       "2        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             374   \n",
       "3        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             376   \n",
       "4        MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...             376   \n",
       "...                                                    ...             ...   \n",
       "1105093  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             381   \n",
       "1105094  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             381   \n",
       "1105095  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             381   \n",
       "1105096  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             381   \n",
       "1105097  MRVMAPQALLLLLSGALALIETWAGSHSMRYFYTAVSRPGRGEPRF...             411   \n",
       "\n",
       "         training  \n",
       "0               1  \n",
       "1               2  \n",
       "2               1  \n",
       "3               1  \n",
       "4               1  \n",
       "...           ...  \n",
       "1105093         2  \n",
       "1105094         1  \n",
       "1105095         1  \n",
       "1105096         1  \n",
       "1105097         2  \n",
       "\n",
       "[1105098 rows x 8 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#dump pickle so I don't have to do previous step again (for now)\n",
    "with open(\"../data/training_samples_5prev.pkl\", 'wb') as f:\n",
    "    pickle.dump(filter_in_df, f)\n",
    "\n",
    "filter_in_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6d49b5bc-6957-424f-ace5-6eee447a059a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tokenizing testing data...\n",
      "Tokenizing training data...\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from transformers import AutoTokenizer\n",
    "\n",
    "training_df = filter_in_df.loc[filter_in_df.training == 1].copy()\n",
    "testing_df = filter_in_df.loc[filter_in_df.training == 2].copy()\n",
    "\n",
    "#Generate data\n",
    "train_sequences = training_df['predict_on'].tolist()\n",
    "train_labels = training_df['label'].tolist()\n",
    "\n",
    "test_sequences = testing_df['predict_on'].tolist()\n",
    "test_labels = testing_df['label'].tolist()\n",
    "\n",
    "#Split data\n",
    "#print(\"Splitting data...\")\n",
    "#train_sequences, test_sequences, train_labels, test_labels = train_test_split(training_sequences, training_labels, test_size=0.25, shuffle=True)\n",
    "\n",
    "del filter_in_df #Hopefully free some memory\n",
    "\n",
    "#Tokenize data\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)\n",
    "\n",
    "print(\"Tokenizing testing data...\")\n",
    "test_tokenized = tokenizer(test_sequences)\n",
    "\n",
    "print(\"Tokenizing training data...\")\n",
    "train_tokenized = tokenizer(train_sequences)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f748bd1d-6d05-4670-bbfd-2a1f460d4866",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../data/training_samples_5prev_traintest_split.pkl\", 'wb') as f:\n",
    "    pickle.dump([train_tokenized, test_tokenized, train_labels, test_labels], f)\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "737ab4b7-c4f2-4f06-b0aa-3af176086d56",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open(\"../data/training_samples_5prev_traintest_split.pkl\", 'rb') as f:\n",
    "    train_tokenized, test_tokenized, train_labels, test_labels = pickle.load(f)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a599fc4e-320b-44df-b80c-4706e1463360",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating test dset...\n",
      "Dumping test dset...\n",
      "Creating train dset...\n",
      "Dumping train dset...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Dataset({\n",
       "    features: ['input_ids', 'attention_mask', 'labels'],\n",
       "    num_rows: 829159\n",
       "})"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Setup for PyTorch\n",
    "from datasets import Dataset\n",
    "\n",
    "print(\"Creating test dset...\")\n",
    "\n",
    "#test_dset = Dataset.from_dict(test_tokenized)\n",
    "#test_dset = test_dset.add_column(\"labels\", test_labels)\n",
    "\n",
    "print(\"Dumping test dset...\")\n",
    "\n",
    "#with open(\"../data/test_dset_prev5.pkl\", 'wb') as f:\n",
    "#    pickle.dump(test_dset, f)\n",
    "\n",
    "del(test_tokenized)\n",
    "del(test_labels)\n",
    "#del(test_dset)\n",
    "\n",
    "print(\"Creating train dset...\")\n",
    "\n",
    "train_dset = Dataset.from_dict(train_tokenized)\n",
    "train_dset = train_dset.add_column(\"labels\", train_labels)\n",
    "\n",
    "print(\"Dumping train dset...\")\n",
    "\n",
    "with open(\"../data/train_dset_prev5.pkl\", 'wb') as f:\n",
    "    pickle.dump(train_dset, f)\n",
    "\n",
    "train_dset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4ad38377-465c-44c1-bc6a-6f3be219b453",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dataset({\n",
       "    features: ['input_ids', 'attention_mask', 'labels'],\n",
       "    num_rows: 828823\n",
       "})"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Let's try training something now... Oh boy\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec06ad51-969a-414d-9b04-c6527e91fe58",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "\"\"\"\n",
    "Create two matrices matrix. For HLA-A, matrix will be 365x30 \n",
    "(length of HLA-AxNumber of HLA-A serotypes -1). We need to \n",
    "identify 30-50 amino acids that exhibit the greatest \n",
    "variety. In this scenario, pairwise comparisons between each\n",
    "2 HLAs, write down proportion of overlap between two\n",
    "\n",
    "While doing the above, create pair-wise vectors, \n",
    "365 long. 0-1 whether identical or not. At the end,\n",
    "sum all 30 vectors to determine sites of interest\n",
    "\"\"\"\n",
    "\n",
    "iterate_alleles = list(allele_mapper.keys())\n",
    "pairwise_vectors = {}\n",
    "\n",
    "for i, allele in enumerate(iterate_alleles):\n",
    "    key_a = allele\n",
    "    for key_b in iterate_alleles:\n",
    "        if key_a[0] != key_b[0]: #Only compare HLA-As to HLA-As, etc\n",
    "            continue\n",
    "        allele_key = \"|\".join(sorted([key_a, key_b]))\n",
    "        if allele_key in pairwise_vectors:\n",
    "            continue\n",
    "    \n",
    "        a1 = allele_mapper[key_a]\n",
    "        a2 = allele_mapper[key_b]\n",
    "\n",
    "        pairwise_vectors[allele_key] = []\n",
    "\n",
    "        len_a1 = len(a1)\n",
    "        len_a2 = len(a2)\n",
    "        use_len = min(len_a1, len_a2)\n",
    "\n",
    "        for idx, aa in enumerate(a1):\n",
    "            if idx < use_len:\n",
    "                if aa == a2[idx]:\n",
    "                    pairwise_vectors[allele_key].append(0)\n",
    "                else:\n",
    "                    pairwise_vectors[allele_key].append(1)\n",
    "\n",
    "        pairwise_vectors[allele_key] = np.array(pairwise_vectors[allele_key])\n",
    "\n",
    "#Get 50-mer index for greatest number of alleles that differ pair-wise\n",
    "A_alleles = []\n",
    "B_alleles = []\n",
    "C_alleles = []\n",
    "G_alleles = []\n",
    "for key, value in pairwise_vectors.items():\n",
    "    if key.startswith('A'):\n",
    "        A_alleles.append(value)\n",
    "    elif key.startswith('B'):\n",
    "        B_alleles.append(value)\n",
    "    elif key.startswith('C'): \n",
    "        C_alleles.append(value)\n",
    "    elif key.startswith('G'):\n",
    "        G_alleles.append(value)\n",
    "\n",
    "A_alleles = np.array(A_alleles)\n",
    "B_alleles = np.array(B_alleles)\n",
    "C_alleles = np.array(C_alleles)\n",
    "G_alleles = np.array(G_alleles) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ffb6ffa-82b1-43f0-b331-67061ae6332c",
   "metadata": {},
   "outputs": [],
   "source": [
    "A_alleles = []\n",
    "B_alleles = []\n",
    "C_alleles = []\n",
    "G_alleles = []\n",
    "for key, value in pairwise_vectors.items():\n",
    "    if key.startswith('A'):\n",
    "        A_alleles.append(value)\n",
    "    elif key.startswith('B'):\n",
    "        B_alleles.append(value)\n",
    "    elif key.startswith('C'): \n",
    "        C_alleles.append(value)\n",
    "    elif key.startswith('G'):\n",
    "        G_alleles.append(value)\n",
    "\n",
    "A_alleles = np.array(A_alleles)\n",
    "B_alleles = np.array(B_alleles)\n",
    "C_alleles = np.array(C_alleles[:-1])\n",
    "G_alleles = np.array(G_alleles)\n",
    "\n",
    "def get_indices_of_highest_values(input_array, return_n_indices):\n",
    "    return(sorted(range(len(input_array)), key = lambda sub: input_array[sub])[-return_n_indices:])\n",
    "\n",
    "A_indices = sorted(get_indices_of_highest_values(sum(A_alleles), 50))\n",
    "B_indices = sorted(get_indices_of_highest_values(sum(B_alleles), 50))\n",
    "C_indices = sorted(get_indices_of_highest_values(sum(C_alleles), 50))\n",
    "\n",
    "#get_indices_of_highest_values(A_alleles, 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7635f9d3-7c5b-4a80-bd98-312bf96dbd17",
   "metadata": {},
   "outputs": [],
   "source": [
    "len([x for x in sum(A_alleles) if x != 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77e0fcac-da29-4833-92a0-b265e1fecf74",
   "metadata": {},
   "outputs": [],
   "source": [
    "len([x for x in sum(B_alleles) if x != 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0a167cf-8c48-4adb-9cb1-b2215f164fac",
   "metadata": {},
   "outputs": [],
   "source": [
    "len([x for x in sum(C_alleles[:-1]) if x != 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16fe17ec-399d-4661-8c26-a75526ed53c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "A_indices = sorted(get_indices_of_highest_values(sum(A_alleles), 50))\n",
    "B_indices = sorted(get_indices_of_highest_values(sum(B_alleles), 50))\n",
    "C_indices = sorted(get_indices_of_highest_values(sum(C_alleles), 50))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
